storage/zynq7000-sdcard/sdcard.c

```diff
@@ -0,0 +1,1212 @@
+/* clang-format on */
+
+
+#define SDHOST_ERROR_REASONS ( \
+	SDHOST_INTR_CMD_ERRORS | \
+	SDHOST_INTR_DAT_ERRORS | \
+	SDHOST_INTR_OVERCURRENT_ERROR | \
+	SDHOST_INTR_AUTO_CMD12_ERROR | \
+	SDHOST_INTR_ADMA_ERROR | \
+	SDHOST_INTR_DMA_ERROR)
+
+#define SDHOST_STATUS_MASK ( \
+	SDHOST_INTR_CMD_STATUS | \
+	SDHOST_INTR_CARD_IN | \
+	SDHOST_INTR_CARD_OUT | \
+	SDHOST_ERROR_REASONS)
+
+/* configuration values */
+#define THREAD_STACK_SIZE 1024
+#define SDHOST_RETRIES    10
+
+/* Number of blocks to erase in a single transaction
+ * Number is limited to not trigger timeouts
+ */
+#define ERASE_N_BLOCKS 128
+
+#define SD_FREQ_INITIAL 400000   /* 400 kHz clock used for card initialization */
+#define SD_FREQ_25M     25000000 /* 25 MHz clock usable when card is initialized */
+#define SD_FREQ_50M     50000000 /* 50 MHz clock usable when card is initialized and supports high speed */
+
+#define ARRAY_LENGTH(array) (sizeof(array) / sizeof(array[0]))
+
+typedef struct {
+	/* Relative Card Address of the card. Needed for selecting the card so it can use the data bus. */
+	uint32_t rca;
+	/* Size of SD card in blocks */
+	uint32_t sizeBlocks;
+	/* Size (in blocks) of an erase sector */
+	uint32_t eraseSizeBlocks;
+	/* Number of command timeouts since last card initialization */
+	uint32_t commandTimeouts;
+	/* Whether the card supports SDHC protocol (requires slightly different handling) */
+	bool highCapacity;
+} sdcard_cardMetadata_t;
+
+typedef struct {
+	/* Base address (in virtual memory) of SD Host Controller register set */
+	volatile uint32_t *base;
+	handle_t cmdLock;
+	sdcard_cardMetadata_t card;
+	/* Reference clock frequency in Hz */
+	uint32_t refclkFrequency;
+
+	/* Address of DMA buffer in virtual memory (for access by the CPU) */
+	void *dmaBuffer;
+	/* Address of DMA buffer in physical memory (for access by the SD Host Controller) */
+	addr_t dmaBufferPhys;
+
+	bool sdioInitialized;
+	bool isCDPinSupported;
+	bool isWPPinSupported;
+
+	handle_t isrHandle;
+	handle_t eventLock;
+	handle_t eventCond;
+} sdcard_hostData_t;
+
+static sdcard_hostData_t sdio_hosts[PLATFORM_SDIO_N_HOSTS] = { 0 };
+static unsigned int initializedHosts = 0;
+
+static struct {
+	handle_t lock;
+	handle_t cond;
+	bool initialized;
+} presenceEvents = { .initialized = false };
+
+static inline sdcard_hostData_t *sdcard_getHostForSlot(unsigned int slot)
+{
+	return (slot >= PLATFORM_SDIO_N_HOSTS) ? NULL : &sdio_hosts[slot];
+}
+
+#define GET_HOST_OR_RETURN(...) \
+	sdcard_hostData_t *host = sdcard_getHostForSlot(slot); \
+	if (host == NULL) { \
+		return __VA_ARGS__; \
+	}
+
+static int sdcard_configClockAndPower(sdcard_hostData_t *host, uint32_t freq);
+static int sdcard_wideAndFast(sdcard_hostData_t *host);
+
+
+/* Resets parts or all of the host according to the given reset type.
+ * Argument must be one of CLOCK_CONTROL_RESET_*
+ */
+static int sdhost_reset(sdcard_hostData_t *host, uint32_t resetType)
+{
+	*(host->base + SDHOST_REG_CLOCK_CONTROL) |= resetType;
```

Comment:
Good place to insert a memory barier
```suggestion
	*(host->base + SDHOST_REG_CLOCK_CONTROL) |= resetType;
	sdcard_dmb();
```
where
```
static void sdcard_dmb(void)
{
	__asm__ volatile ("dmb");
}
```

Reply:
I inserted a memory barrier there and also here: https://github.com/phoenix-rtos/phoenix-rtos-devices/blob/2ef22535ad0bfe6a556013d37ac638bf443160ac/storage/zynq7000-sdcard/sdcard.c#L897-L898